### PR TITLE
[FSDP] Add HSDP parity unit test

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106080
* #106068
* #106034
* __->__ #106131
* #106072
* #106033
* #106032

With >= 4 GPUs:
```
python -m pytest test/distributed/fsdp/test_fsdp_hybrid_shard.py -k test_fsdp_hybrid_shard_parity
```

Differential Revision: [D47852458](https://our.internmc.facebook.com/intern/diff/D47852458)

### CODE REVIEW

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47852458).

File: test/distributed/fsdp/test_fsdp_hybrid_shard.py [hunk 1]
```diff
@@ -313,11 +317,73 @@ def patched_collective(orig_collective, counter, *args, **kwargs):
...
+            use_orig_params,
+            hsdp_process_groups=hsdp_pgs,
+        )
+        assert (
+            hsdp_model._inter_node_pg.size() > 1
+        ), "HSDP model initialized without replication"
+        fsdp_optim = torch.optim.Adam(fsdp_model.parameters(), lr=1e-2)
+        hsdp_optim = torch.optim.Adam(hsdp_model.parameters(), lr=1e-2)
+        for _ in range(5):
+            inp = fsdp_model.module.get_input(torch.device("cuda"))
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_hybrid_shard.py:351]: you might have to seed differently across ranks to ensure that inputs are diff across ranks for this test to be nontrivial?

**Review by rohan-varma — APPROVED** [review]

File: test/distributed/fsdp/test_fsdp_hybrid_shard.py (hunk 1 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_hybrid_shard.py:352]: Fixed!
